impl/src/main/java/org/jboss/weld/injection/producer/ProducerMethodProducer.java

```diff
@@ -69,22 +71,30 @@ protected void checkProducerMethod(EnhancedAnnotatedMethod<T, ? super X> method)
             boolean methodDeclaredOnTypes = false;
             // TODO use annotated item?
             for (Type type : getDeclaringBean().getTypes()) {
-                // TODO: deal with parameterized types!
-                if (type instanceof Class<?>) {
-                    try {
-                        AccessController.doPrivileged(new GetMethodAction((Class<?>) type, method.getName(), method.getParameterTypesAsArray()));
-                        methodDeclaredOnTypes = true;
-                        break;
-                    } catch (PrivilegedActionException ignored) {
-                    }
+                Class<?> clazz = Reflections.getRawType(type);
+                methodDeclaredOnTypes = doPrivileged(clazz, method.getName(), method.getParameterTypesAsArray());
+                if (methodDeclaredOnTypes) {
+                    break;
                 }
             }
+            if (!methodDeclaredOnTypes && Modifier.isStatic(method.slim().getJavaMember().getModifiers())) {
```

Comment:
If the method is static then we should not perform this check at all, since the spec only covers non-static session bean classes. Therefore, Modifier.isStatic() should be moved upwards to the outer if condition.